src/tests/tools/iotc_test_echoserver.cc

```diff
@@ -14,72 +92,63 @@ uint16_t EchoTestServer::CreateServer() {
   sprintf(port_s, "%d", test_port_);
 
   memset(&hints, 0, sizeof(struct addrinfo));
-  hints.ai_family = AF_UNSPEC;
-  hints.ai_socktype = sock_type_;
+  hints.ai_family = protocol_type_;
+  hints.ai_socktype = socket_type_;
   hints.ai_flags = AI_PASSIVE;
   status = getaddrinfo(NULL, port_s, &hints, &result);
   if (0 != status) {
-    return 1;
+    return ServerError::kFailedGetAddrInfo;
   }
 
   for (rp = result; rp != NULL; rp = rp->ai_next) {
-    if ((server_sock_ =
+    if ((server_socket_ =
              socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol)) < 0) {
       continue;
     }
-    printf("protocol type: %d\n", rp->ai_family);
     int reuseAddress = 1;
-    if (setsockopt(server_sock_, SOL_SOCKET, SO_REUSEADDR, &reuseAddress,
+    if (setsockopt(server_socket_, SOL_SOCKET, SO_REUSEADDR, &reuseAddress,
                    sizeof(reuseAddress))) {
-      perror("[Server] Error setting socket reuseaddr option");
-      close(server_sock_);
-      return 1;
+      close(server_socket_);
+      return ServerError::kFailedSetSockOpt;
     }
-    if (bind(server_sock_, rp->ai_addr, rp->ai_addrlen) == 0)
+    if (bind(server_socket_, rp->ai_addr, rp->ai_addrlen) == 0)
       break;
     else
-      close(server_sock_);
+      close(server_socket_);
   }
 
   if (rp == NULL) {
-    perror("Error binding");
-    return 1;
+    return ServerError::kError;
   }
   freeaddrinfo(result);
 
   tv.tv_sec = kTimeoutSeconds;
   tv.tv_usec = 0;
-  if (setsockopt(server_sock_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) < 0) {
-    perror("[Server] Error setting socket timeout option");
-    close(server_sock_);
-    return 1;
+  if (setsockopt(server_socket_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) <
+      0) {
+    close(server_socket_);
+    return ServerError::kFailedSetSockOpt;
   }
-  if (sock_type_ == SOCK_STREAM) {
-    if (listen(server_sock_, 5) < 0) {
-      perror("[Server] Error listening connection");
-      return 1;
+  if (socket_type_ == SOCK_STREAM) {
+    if (listen(server_socket_, 5) < 0) {
+      return ServerError::kFailedListen;
     }
   }
 
-  return 0;
+  return ServerError::kSuccess;
 }
 
-uint16_t EchoTestServer::RunServer() {
-  if (sock_type_ == SOCK_STREAM) {
-    if (RunTcpServer() != 0) {
-      return 1;
-    }
-  } else if (sock_type_ == SOCK_DGRAM) {
-    if (RunUdpServer() != 0) {
-      return 1;
-    }
+EchoTestServer::ServerError EchoTestServer::RunServer() {
+  if (socket_type_ == SOCK_STREAM) {
+    return RunTcpServer();
+  } else if (socket_type_ == SOCK_DGRAM) {
+    return RunUdpServer();
   }
-  return 0;
+  return ServerError::kSuccess;
 }
 
-char* EchoTestServer::get_recv_buf() { return recv_buf_; }
-
-void EchoTestServer::StopServer() {
+void EchoTestServer::Stop() {
```

Comment:
Here, you would need to terminate the server thread. After calling `Stop`, the server thread may be in a blocking state - e.g. in `accept()`.

Use an `std::unique_ptr<std::thread> thread_;` and here, you could just `thread_.reset(nullptr);` or `delete thread_.release();`